build 0.10.0

h-vetinari opened the pull request:
Try building https://github.com/serge-sans-paille/pythran/pull/1894

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
Seems that the xsimd-unvendoring for windows hasn't worked for a while... (not least because the line `- python -c "import os, shutil; shutil.rmtree(os.path.join('pythran', 'xsimd'))"` erroring did not cause the build to stop).

The xsimd package is in the env, the headers are [there](https://github.com/conda-forge/xsimd-feedstock/blob/master/recipe/meta.yaml#L27), the `-I%PREFIX%\include` is there, so I really don't understand why `#include <xsimd/xsimd.hpp>` fails.

serge-sans-paille commented:
@h-vetinari can you apply the following patch so that we get log info:

```
diff --git a/omp/__init__.py b/omp/__init__.py
index 3496891..6561f6a 100644
--- a/omp/__init__.py
+++ b/omp/__init__.py
@@ -66,27 +66,36 @@ class OpenMP(object):
             if env_paths:
                 paths.extend(env_paths.split(os.pathsep))
 
+        print("paths:", paths)
+
         libomp_names = self.get_libomp_names()
+
+        print("libomp_names:", libomp_names)
         for libomp_name in libomp_names:
 
             cmd = [cxx, '-print-file-name=' + libomp_name]
             # The subprocess can fail in various ways, including because it
             # doesn't support '-print-file-name'. In that case just give up.
             try:
+                print("running:", cmd)
                 path = os.path.dirname(check_output(cmd).decode().strip())
                 if path:
                     paths.append(path)
             except (OSError, CalledProcessError):
                 pass
 
+        print("paths:", paths)
+
         for libomp_name in libomp_names:
             # Try to load find libomp shared library using loader search dirs
             libomp_path = find_library(libomp_name)
+            print(libomp_name, "->", libomp_path)
 
             # Try to use custom paths if lookup failed
             if not libomp_path:
                 for path in paths:
                     candidate_path = os.path.join(path, libomp_name)
+                    print("trying:", candidate_path)
                     if os.path.isfile(candidate_path):
                         libomp_path = candidate_path
                         break
```

serge-sans-paille commented:
mmh ``x86_64-apple-darwin13.4.0-clang++ -print-file-name=libiomp5.dylib`` seems to fail. Do you have a way to check that?

h-vetinari commented:
Shouldn't we be expecting to [find](https://github.com/conda-forge/openmp-feedstock/blob/master/recipe/meta.yaml#L53) `libomp.dylib` (since llvm-openmp is in the environment)?

isuruf commented:
That's expected as the clang doesn't really have a standard place for libomp.dylib.

What can be fixed in pythran is to fix the call to `find_library`. `find_library` is called with `libomp.dylib`, but from docs,

> ctypes.util.find_library(name)

> Try to find a library and return a pathname. name is the library name without any prefix like lib, suffix like .so, .dylib or version number (this is the form used for the posix linker option -l). If no library can be found, returns None.

isuruf commented on recipe/patches/0001-0.10.0-palig.patch:
```diff
@@ -0,0 +1,3230 @@
+-
++        """Return list of OpenMP libraries to try"""
++        names = ['omp', 'gomp', 'iomp5']
+         return [name + get_shared_lib_extension() for name in names]
```
you need to remove the shared lib extension too

serge-sans-paille replied:
ok, and use the extension only when needed.

h-vetinari commented:
Thanks a lot @isuruf! 😊

Any idea why the xsimd-unvendoring on windows is not working despite the [having](https://github.com/conda-forge/pythran-feedstock/pull/56#issuecomment-915917812) the lib, the headers, the `-I` include...?

Summary of windows build status changes...
| commit | status |
|-|-|
| https://github.com/conda-forge/pythran-feedstock/pull/56/commits/fa920dae340c2003867ce3636d2801ac5a2b6237 | ✔️ |
| https://github.com/conda-forge/pythran-feedstock/pull/56/commits/ed65adcd57f2b0b1ebf0562ee1d9fc9d29e7f397| ❌  |
| https://github.com/conda-forge/pythran-feedstock/pull/56/commits/90703abe7f332241ef3f470f53412b7010cafce1 | ✔️ |
| https://github.com/conda-forge/pythran-feedstock/pull/56/commits/a35ed2158adeed478957da82454230df091c1b7d| ❌  |

isuruf commented:
You probably need to set `compiler.include_dirs = %LIBRARY_PREFIX%\include` in setup.cfg

isuruf commented:
see https://github.com/serge-sans-paille/pythran/blob/f5832d4eb05ebcda7598b507fee6a1bf1b503716/pythran/pythran-win32.cfg#L4

h-vetinari commented:
> 
> 
> You probably need to set `compiler.include_dirs = %LIBRARY_PREFIX%\include` in setup.cfg

Ah, now I think I get it. I had seen the  `-I%PREFIX%\include` and forgotten that this is actually the wrong path on windows, cf. https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html?highlight=cmake#environment-variables-set-during-the-build-process

serge-sans-paille commented:
I'll apply the omp patch upstream.

serge-sans-paille commented:
upstream branch updated!

h-vetinari commented:
> 
> 
> You probably need to set `compiler.include_dirs = %LIBRARY_PREFIX%\include` in setup.cfg

I mean, I _can_ patch this into windows builds, but it sounds to me like the include_dirs is wrongly populated (with `%PREFIX%\include` instead of `%PREFIX%\Library\include`) somewhere on the way from conda to the pythran-setup (and I couldn't quickly [unravel](https://github.com/serge-sans-paille/pythran/search?q=include_dirs) it on the pythran side). Wouldn't this be better to fix without a patch, if possible?

serge-sans-paille commented:
Let's patch it at the conda level, then open a PR on pythran that will land in next release?

serge-sans-paille commented:
sorry, you picked a trnasinet version, I updated the upstream branch once more :-/

h-vetinari commented:
> 
> 
> sorry, you picked a trnasinet version, I updated the upstream branch once more :-/

No, I got the most recent one, but made a mistake with the backslash-handling...

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found some lint.

Here's what I've got...


For **recipe**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

isuruf commented on recipe/meta.yaml:
```diff
@@ -1,22 +1,32 @@
+    # since this is executed through batch, dealing with '%' is horrible, especially because also cfg-parsing
+    # (see docs.python.org/3/library/configparser.html) would fail on bare % even if we manage to write correctly;
+    # avoid all the absurd consumption rules of '%' in batch by simply doing % -> chr(37)
+    - "python -c \"with open('setup.cfg', 'a') as s: s.write(f'[compiler]\\ninclude_dirs={chr(37)*2}LIBRARY_PREFIX{chr(37)*2}\\include')\""  # [win]
```
Why are you trying to have `%` in the file? We should expand the env variable before writing to the file

h-vetinari replied:
Disagree. Writing `%LIBRARY_PREFIX%` turns into `%PREFIX%\Library` turns into `D:\bld\...` which gets written as `D:ld\...` (because python interprets `\b`).

In other words, all the same escaping problems remain for writing the expanded variable, and after lots of futile tries, I'm pretty sure the `chr(37)` solution is the most elegant one (including legibility and reasonable length).

h-vetinari replied:
Though, on second though, I'll try again with `r'...'` and see what comes out.

h-vetinari replied:
... which - in turn - makes it hard to write line breaks into the cfg... All a bit tricky (without direct patching, which I'd like to avoid, however) 🙃

But open for ideas (that work 😄)

h-vetinari replied:
OK, I'd have a solution** but then the issue is (since changed in the code) - we're not just talking about `setup.cfg` anymore (which wasn't active/effective at test time), but about the `pythran.cfg` that'll be part of the distributed artefact. So I think we have no choice but to write with `%`.

**
```diff
     # on windows, pythran wrongly sets %PREFIX\include instead of %PREFIX%\Library\include == %LIBRARY_PREFIX%\include;
-    # since this is executed through batch, dealing with '%' is horrible, especially because also cfg-parsing
-    # (see docs.python.org/3/library/configparser.html) would fail on bare % even if we manage to write correctly;
-    # avoid all the absurd consumption rules of '%' in batch by simply doing % -> chr(37)
-    - "python -c \"with open('pythran\\pythran.cfg', 'a') as s: s.write(f'\\n[compiler]\\ninclude_dirs={chr(37)*2}LIBRARY_PREFIX{chr(37)*2}\\include')\""  # [win]
+    # need to write line breaks '\n' as well as escape %LIBRARY_PREFIX% with r'...' -> separate strings
+    - "python -c \"with open('pythran\\pythran.cfg', 'a') as s: s.write('\\n[compiler]\\ninclude_dirs=' + r'%LIBRARY_PREFIX%\include')\""  # [win]
     # sanity check to ensure this is set correctly
     - ```

isuruf replied:
> , but about the pythran.cfg that'll be part of the distributed artefact. So I think we have no choice but to write with %.

why not? conda will fix the prefix in the distributed artefact.

h-vetinari replied:
Maybe I don't understand you correctly - I thought you meant writing the final value (something like `D:\bld\pythran_1631295533774\work`), which is obviously not portable.

Are you saying it should be
```
[compiler]
include_dirs=%%PREFIX%%\Library\include
```

Then I'm completely fine with that (but it still has to escape `%`)...

isuruf replied:
> Maybe I don't understand you correctly - I thought you meant writing the final value (something like D:\bld\pythran_1631295533774\work), which is obviously not portable.

Yes, I meant writing the final value like `D:\bld\pythran_1631295533774\_h_env`.
Why is it obviously not portable?

h-vetinari replied:
> Why is it obviously not portable?

Is this being caught by conda and replaced by `%PREFIX%` (in a random cfg-file)? Then I can believe it, I just didn't think of that.

h-vetinari replied:
> in a random cfg-file

Conda would also have to be smart enough to write it as `%%PREFIX%%`, otherwise the cfg-parsing breaks...

isuruf replied:
> Is this being caught by conda 

yes

> and replaced by %PREFIX% (in a random cfg-file)? Then I can believe it, I just didn't think of that.

No, it would be replaced by the correct value at install time in the user's machine.

h-vetinari replied:
> No, it would be replaced by the correct value at install time in the user's machine.

That sounds very magical to me, but OK, cool.

h-vetinari replied:
For some reason `%PREFIX%` doesn't seem to be set now...?

Despite inserting it without any escapes, the result of `type pythran\pythran.cfg` is:
```
[...snip....]

[compiler]
include_dirs=%PREFIX%\Library\include
```

h-vetinari commented:
> 
> 
> Let's patch it at the conda level, then open a PR on pythran that will land in next release?

Well, I think I found a way to do this without patching, but whether I add 
```
[compiler]
include_dirs=%%LIBRARY_PREFIX%%\include
```

to `setup.cfg` _or_  to `pythran\pythran.cfg`, it does not get picked up in the final binary.

Need your help on that please, @serge-sans-paille.

Also, as noted in the upstream issue, the new openmp setup on windows is very [noisy](https://github.com/serge-sans-paille/pythran/pull/1894#issuecomment-917071392).

serge-sans-paille commented:
@h-vetinari : I think  you should update the *source* ``pythran/pythran-win32.cfg`` before install

h-vetinari commented:
> 
> 
> @h-vetinari : I think you should update the _source_ `pythran/pythran-win32.cfg` before install

Right now I'm changing the `pythran\pythran.cfg` (but only on windows) - is that functionally any different?

h-vetinari commented:
@serge-sans-paille, I'm now showing the `pythran.cfg` from the site-packages directory (during the test phase), i.e. `type %PREFIX%\Lib\site-packages\pythran\pythran.cfg`:
which now contains
```
[...snip...]

[compiler]
include_dirs=D:/bld/pythran_1631305376390/_test_env\Library\include
```

However, this does not seem to get inserted anywhere in the actual calls...?

isuruf commented:
As we both said, you need to patch `pythran-win32.cfg`.

serge-sans-paille commented on recipe/meta.yaml:
```diff
@@ -1,22 +1,31 @@
+    # on windows, pythran wrongly sets %PREFIX\include instead of %PREFIX%\Library\include;
+    # need to write line breaks '\n' as well as escape %PREFIX% with r'...' -> separate strings;
+    # conda will detect the environment-path here and replace it as appropriate on user installs
+    - "python -c \"with open('pythran\\pythran.cfg', 'a') as s: s.write('\\n[compiler]\\ninclude_dirs=' + r'%PREFIX%\\Library\\include')\""  # [win]
```
should be ``'pythran\\pythran-win32.cfg'`` here

h-vetinari commented:
Finally, this passes without vendored xsimd also on windows 🥳 

From my POV, this is now ready (well, except for the upstream release of 0.10.0 and rebasing).

BTW, I'd be willing to join as maintainer @conda-forge/pythran

h-vetinari commented:
Finally, this passes without vendored xsimd also on windows 🥳 

From my POV, this is now ready (well, except for the upstream release of 0.10.0 and rebasing).

BTW, I'd be willing to join as maintainer @conda-forge/pythran

h-vetinari commented:
> 
> 
> Finally, this passes without vendored xsimd also on windows 🥳

Thanks for the help & patience @isuruf & @serge-sans-paille!

h-vetinari commented:
Interestingly, there's one error I hadn't seen yet (due to the quicker iteration cycle) - pypy on aarch complains:
```
Successfully built pythran
Installing collected packages: pythran
ERROR: Exception:
Traceback (most recent call last):
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/cli/base_command.py", line 173, in _main
    status = self.run(options, args)
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/cli/req_command.py", line 203, in wrapper
    return func(self, options, args)
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/commands/install.py", line 399, in run
    pycompile=options.compile,
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/req/__init__.py", line 81, in install_given_reqs
    pycompile=pycompile,
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/req/req_install.py", line 766, in install
    requested=self.user_supplied,
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/operations/install/wheel.py", line 802, in install_wheel
    requested=requested,
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/operations/install/wheel.py", line 637, in _install_wheel
    file.save()
  File "/drone/src/build_artifacts/[...]/site-packages/pip/_internal/operations/install/wheel.py", line 405, in save
    with open(self.dest_path, "wb") as dest:
ValueError: embedded null byte
```

serge-sans-paille commented:
It's difficult for me to make a connection with pythran there :-)

h-vetinari commented:
@mattip
Have you encountered `ValueError: embedded null byte` with pypy on aarch somewhere recently...?

mattip commented:
I have not seen that. It seems something is off with the file name:
```
...
site-packages/pip/_internal/operations/install/wheel.py", line 405, in save
    with open(self.dest_path, "wb") as dest:
ValueError: embedded null byte
```

h-vetinari commented:
> I have not seen that. It seems something is off with the file name:

Yeah, that's my suspicion as well, but it only happens for pypy on aarch, which is weird.

h-vetinari commented:
Looking at the previous [runs](https://cloud.drone.io/conda-forge/pythran-feedstock/), it could also just be a flake... Perhaps not worth to invest much time here for the moment.

h-vetinari commented:
BTW, pypy failure on arch didn't show up again. :)

h-vetinari commented:
Rebased.

> BTW, I'd be willing to join as maintainer @conda-forge/pythran

As suggested further up, I've now added myself as a maintainer. Let me know if that's not desirable and I'll remove the last commit.

serge-sans-paille commented:
@h-vetinari Indeed. I merged #57 because it was passing validation, thinking it doesn't prevent us from rebasing this PR on it. But it does involve extra work, so I'm pretty sorry for that :-/

h-vetinari commented:
It's less about a bit of rebasing and more about avoiding to publish subpar builds (because, despite the green CI the problems found & fixed in this PR were obviously not solved in #57) pointlessly.

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -5,18 +5,26 @@ package:
 source:
-  url: https://pypi.io/packages/source/p/pythran/pythran-{{ version }}.tar.gz
-  sha256: 9dac8e1d50f33d4676003e350b1f0c878ce113e6f907920e92dc103352cac5bf
+  url: https://github.com/serge-sans-paille/pythran/archive/refs/tags/{{ version }}.tar.gz
+  sha256: c4853430e99dc382b8ed11e6eed3a196a571b2f311db4be4cf7b28d6f6866c54
```
This is not strictly speaking necessary, but allows easier patchability (there were some issues using PyPI sources). It's also more transparent, IMO.

h-vetinari commented:
Ping 🙃

h-vetinari commented:
Ping @serge-sans-paille

serge-sans-paille commented:
Thanks for the hard work, and... sorry for the delay. wow, 9 days since last ping :-/
